base/fastmath.jl

```diff
@@ -0,0 +1,329 @@
+# Support for @fastmath
+
+# This module provides versions of math functions that may violate
+# strict IEEE semantics.
+
+# This allows the following transformations:
+# nnan: No NaNs - Allow optimizations to assume the arguments and
+#       result are not NaN. Such optimizations are required to retain
+#       defined behavior over NaNs, but the value of the result is
+#       undefined.
+# ninf: No Infs - Allow optimizations to assume the arguments and
+#       result are not +/-Inf. Such optimizations are required to
+#       retain defined behavior over +/-Inf, but the value of the
+#       result is undefined.
+# nsz:  No Signed Zeros - Allow optimizations to treat the sign of a
+#       zero argument or result as insignificant.
+# arcp: Allow Reciprocal - Allow optimizations to use the reciprocal
+#       of an argument rather than perform division.
+
+module FastMath
+
+export @fastmath
+
+import Core.Intrinsics: box, unbox, powi_llvm, sqrt_llvm_fast
+
+const fast_op =
+    Dict(# basic arithmetic
+         :+ => :add_fast,
+         :- => :sub_fast,
+         :* => :mul_fast,
+         :/ => :div_fast,
+         :(==) => :eq_fast,
+         :!= => :ne_fast,
+         :< => :lt_fast,
+         :<= => :le_fast,
+         :abs => :abs_fast,
+         :abs2 => :abs2_fast,
+         :cmp => :cmp_fast,
+         :conj => :conj_fast,
+         :inv => :inv_fast,
+         :mod => :mod_fast,
+         :rem => :rem_fast,
+         :sign => :sign_fast,
+         :isfinite => :isfinite_fast,
+         :isinf => :isinf_fast,
+         :isnan => :isnan_fast,
+         :issubnormal => :issubnormal_fast,
+         # math functions
+         :^ => :pow_fast,
+         :acos => :acos_fast,
+         :acosh => :acosh_fast,
+         :angle => :angle_fast,
+         :asin => :asin_fast,
+         :asinh => :asinh_fast,
+         :atan => :atan_fast,
+         :atan2 => :atan2_fast,
+         :atanh => :atanh_fast,
+         :cbrt => :cbrt_fast,
+         :cis => :cos_fast,
+         :cos => :cos_fast,
+         :cosh => :cosh_fast,
+         :exp10 => :exp10_fast,
+         :exp2 => :exp2_fast,
+         :exp => :exp_fast,
+         :expm1 => :expm1_fast,
+         :hypot => :hypot_fast,
+         :lgamma => :lgamma_fast,
+         :log10 => :log10_fast,
+         :log1p => :log1p_fast,
+         :log2 => :log2_fast,
+         :log => :log_fast,
+         :max => :max_fast,
+         :min => :min_fast,
+         :minmax => :minmax_fast,
+         :sin => :sin_fast,
+         :sinh => :sinh_fast,
+         :sqrt => :sqrt_fast,
+         :tan => :tan_fast,
+         :tanh => :tanh_fast)
```

Comment:
As an alternative to using a dictionary: since all of these functions are simply mapped from `fn` to `fn_fast`, you *might* consider making `FastMath` a `baremodule` (so that nothing is imported by default), and simply redefine the functions using the original function names.  Then the fast functions could be referred to using `FastMath.fn`.  (You also wouldn't import all of them in `sysimg.jl` below, of course--probably just the `@fastmath` macro.)

I guess you might still need/want a list of functions that have `FastMath` variants, to figure out what to rewrite and what to ignore in the macro....